internal/cli/server/server.go

```diff
@@ -50,20 +50,40 @@ type Server struct {
 	tracerAPI *tracers.API
 }
 
-func NewServer(config *Config) (*Server, error) {
+type serverOption func(srv *Server, config *Config) error
+
+func WithGRPCAddress() serverOption {
+	return func(srv *Server, config *Config) error {
+		return srv.gRPCServerByAddress(config.GRPC.Addr)
+	}
+}
+
+func WithGRPCListener(lis net.Listener) serverOption {
+	return func(srv *Server, _ *Config) error {
+		return srv.gRPCServerByListener(lis)
+	}
+}
+
+//nolint:gocognit
+func NewServer(config *Config, opts ...serverOption) (*Server, error) {
 	srv := &Server{
 		config: config,
 	}
 
 	// start the logger
 	setupLogger(config.LogLevel)
 
-	if err := srv.setupGRPCServer(config.GRPC.Addr); err != nil {
-		return nil, err
+	var err error
+
+	for _, opt := range opts {
```

Comment:
Should we handle the case where `len(opts) = 0`?

Reply:
nope )

Reply:
it'll skip it safely if len(opts) == 0

Reply:
But then it won't start grpc server, right?

Reply:
if opts is empty then we just don't have to run grpc server.